lib/IRGen/IRGenSIL.cpp

```diff
@@ -2074,31 +2153,62 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
       assert(0 && "TensorAttributes cannot be dynamic");
     case GraphOperationInfo::ArgumentLowering::Input: {
       assert(argumentName.empty() && "inputs cannot have names");
+
+      // Keep track of how many inputs we have added for this
+      // `structuredArgument`. (It is possible to add more than one input per
+      // `structuredArgument` in the case of an input list. See the
+      // "ParseExample" example above for more information.)
+      llvm::Value *listLength = llvm::ConstantInt::get(IGM.Int32Ty, 0);
+      auto incrementListLength = [&](llvm::Value *amount) {
+        listLength = Builder.CreateAdd(listLength, amount);
+      };
+
       switch (structuredArgument.getKind()) {
       case GraphOperationInfo::SAK_Single: {
         LLVM_DEBUG(llvm::dbgs() << " Adding a single input.\n");
         auto tensorHandle = structuredArgument.getSingleArgument();
-        addTensorhandleAsOpInput(tensorHandle);
+        incrementListLength(unpackAndAddInput(tensorHandle));
```

Comment:
i still find it a bit unintuitive and complex that we are calling `incrementListLength()` in this context, but there is no loop around it. It'd probably be cleaner if we instead call another function here, sth like `setListLength()`. this way we can probably also move `llvm::Value *listLength = llvm::ConstantInt::get(IGM.Int32Ty, 0);` to the `case GraphOperationInfo::SAK_List` block below. 

But my preference is not very strong here.

Reply:
I changed it to set it directly in the non-loop case

Reply:
love the new code structure!